crowbar_framework/app/models/api/upgrade.rb

```diff
@@ -24,6 +24,10 @@ def status
       end
 
       def checks
+        upgrade_status = ::Crowbar::UpgradeStatus.new
+        # the check for current_step means to allow running the step at any point in time
+        upgrade_status.start_step if upgrade_status.current_step == "upgrade_prechecks"
```

Comment:
This causes a race window to open in between the equality check and the call to `#start_step`.  This is almost exactly the same race condition I was predicting in [my earlier comment](https://github.com/crowbar/crowbar-core/pull/692#issuecomment-252298193):

> Also it seems strange that `#next_step!` doesn't return the next step. This means that the consumer has to call `#current_step` immediately after, which feels like it's introducing another race window in between advancing the step and finding out what the new step is in order to be able to start executing it.

The only difference is that here you are querying the current step *before* marking it as started, instead of after.

(Actually, there is another worse problem, which is that `current_step` returns the in-memory state which is not protected by a lock, and is consequently not thread-safe.  But we can fix that pretty easily by changing `#current_step` to always read from the JSON file and not cache in-memory.)

To do this correctly, the locking library needs to provide [an atomic test-and-set](https://en.wikipedia.org/wiki/Test-and-set) operation.  For example, the interface could be changed so that the caller (here) would do:

    upgrade_status.start_step("upgrade_prechecks")

and then the test and set would both happen inside the critical region protected by a single lock.  If the current step isn't `upgrade_prechecks` then this call would either raise an exception or return `nil` or `false` or something like that.